unixfs/src/file.rs

```diff
@@ -0,0 +1,479 @@
+                hash_type, fanout
+            ),
+        }
+    }
+}
+
+impl std::error::Error for FileError {}
+
+impl From<FileError> for FileReadFailed {
+    fn from(e: FileError) -> Self {
+        Self::File(e)
+    }
+}
+
+/// This exists to help matching the borrowed slice in `Option<Cow<'_, [u8]>>` in this file
+/// or defaulting to empty array. In the processing inside this file, the Cow never represents
+/// owned value.
+///
+/// This at least sounded useful early on as the quick-protobuf produces many Option<Cow> values
+/// which are a bit tricky to handle. We never turn them into Option<Cow::Owned> so we can safely
+/// use these.
+pub(crate) trait UnwrapBorrowedExt<'a> {
+    /// Does not default but requires there to be an borrowed inner value.
+    fn unwrap_borrowed(self) -> &'a [u8];
+
+    /// Unwraps the Cow of [u8] into empty or wrapped slice.
+    fn unwrap_borrowed_or_empty(self) -> &'a [u8]
+    where
+        Self: 'a;
+}
+
+impl<'a> UnwrapBorrowedExt<'a> for Option<Cow<'a, [u8]>> {
+    fn unwrap_borrowed(self) -> &'a [u8] {
+        match self {
+            Some(Cow::Borrowed(x)) => x,
+            Some(Cow::Owned(_)) => panic!("unexpected Cow::Owned"),
+            None => panic!("Unexpected None"),
+        }
+    }
+    fn unwrap_borrowed_or_empty(self) -> &'a [u8] {
+        match self {
+            Some(Cow::Borrowed(x)) => x,
+            None => &[][..],
+            _ => panic!("should not be Cow::Owned"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{reader::*, visit::*, UnwrapBorrowedExt};
+    use hex_literal::hex;
+    use std::convert::TryFrom;
+
+    const CONTENT_FILE: &'static [u8] = &[
+        0x0a, 0x0d, 0x08, 0x02, 0x12, 0x07, 0x63, 0x6f, 0x6e, 0x74, 0x65, 0x6e, 0x74, 0x18, 0x07,
+    ];
+
+    #[test]
+    fn just_content() {
+        let fr = FileReader::from_block(CONTENT_FILE).unwrap();
+
+        match fr.content() {
+            (FileContent::Just(b"content"), _) => {}
+            x => panic!("unexpected: {:?}", x),
+        }
+    }
+
+    #[test]
+    fn visiting_just_content() {
+        let res = IdleFileVisit::default().start(CONTENT_FILE);
+
+        match res {
+            Ok((b"content", _, None)) => {}
+            x => unreachable!("unexpected {:?}", x),
+        }
+    }
+
+    #[test]
+    fn visiting_too_large_range_of_singleblock_file() {
+        let res = IdleFileVisit::default()
+            .with_target_range(500_000..600_000)
+            .start(CONTENT_FILE);
+
+        match res {
+            Ok((b"", _, None)) => {}
+            x => unreachable!("unexpected {:?}", x),
+        }
+    }
+
+    #[test]
+    fn empty_file() {
+        let block = &[0x0a, 0x04, 0x08, 0x02, 0x18, 0x00];
+        let fr = FileReader::from_block(block).unwrap();
+        assert_eq!(fr.content().0.unwrap_content(), &[][..]);
+    }
+
+    use cid::Cid;
+    use std::collections::HashMap;
```

Comment:
nit: those imports could accompany the others at the top of the module

Reply:
Moved